File: android/app/src/main/java/com/pan3d/md5/MeshToObjUtils.java
package com.pan3d.md5;

import com.pan3d.base.MeshData;
import com.pan3d.md5.vo.Md5MeshData;
import com.pan3d.md5.vo.MeshItem;
import com.pan3d.md5.vo.ObjectBone;
import com.pan3d.md5.vo.ObjectUv;
import com.pan3d.md5.vo.ObjectWeight;
import com.pan3d.vo.Matrix3D;
import com.pan3d.vo.Quaternion;
import com.pan3d.vo.Vector3D;

import java.util.ArrayList;
import java.util.List;

public class MeshToObjUtils {
    public static List<ObjectBone> getStorNewTargerArr(List<ObjectBone> targetAry) {
        List<ObjectBone> newTargetAry = new ArrayList<>();
        for (int i = 0; i < targetAry.size(); i++) {
            if (newTargetAry.indexOf(targetAry.get(i)) == -1) {
                newTargetAry.add(targetAry.get(i));
            }
        }
        return newTargetAry;
    }

    public MeshData getObj(Md5MeshData mesh) {
        MeshData objData = mesh;
        objData.verticeslist = new ArrayList<>();
        objData.uvlist = new ArrayList<>();
        objData.normals = new ArrayList<>();
        objData.indexs = new ArrayList<>();
        List<Matrix3D> bindPosAry  = new ArrayList<>();
        List<Matrix3D>  invertAry = new ArrayList<>();
        List<MeshItem> meshItemAry = new ArrayList<>();
        List<ObjectBone> boneItemAry = this.processBoneNew(mesh.boneItem);
        for (int i = 0; i < boneItemAry.size(); i++) {
            ObjectBone objbone = boneItemAry.get(i);
            Quaternion OldQ = new Quaternion(objbone.qx, objbone.qy, objbone.qz);
            OldQ.w = this.getW(OldQ.x, OldQ.y, OldQ.z);
            Matrix3D newM = OldQ.toMatrix3D();
            newM.appendTranslation(objbone.tx, objbone.ty, objbone.tz);
            objbone.matrix = newM;
            bindPosAry.add(newM);
            Matrix3D inverMatrix = newM.clone();
            inverMatrix.invert();
            invertAry.add(inverMatrix);
        }

        for (int i = 0; i < mesh.uvItem.size(); i++) {
            ObjectUv objuv = mesh.uvItem.get(i);
            Vector3D v3d  = new Vector3D();
            List<Float> wAry = new ArrayList<>();
            for (int j = 0; j < objuv.b; j++) {
                int weightID = objuv.a + j;
                ObjectWeight objWeight = mesh.weightItem.get(weightID);
                Matrix3D ma = boneItemAry.get(objWeight.boneId).matrix;

                Vector3D tempV3d = new Vector3D(objWeight.x, objWeight.y, objWeight.z);
                tempV3d = ma.transformVector(tempV3d);
                tempV3d.scaleBy(objWeight.w);
                v3d = v3d.add(tempV3d);
                wAry.add(objWeight.w);
            }
            objData.verticeslist.add(v3d.x);
            objData.verticeslist.add(v3d.y);
            objData.verticeslist.add(v3d.z);
            objData.uvlist.add(objuv.x);
            objData.uvlist.add(objuv.y);
            MeshItem meshitem = new MeshItem();
            meshitem.verts = new Vector3D(v3d.x, v3d.y, v3d.z);
            meshitem.uvInfo = objuv;
            meshItemAry.add(meshitem);
        }
        for (int i = 0; i < mesh.triItem.size(); i++) {
            objData.indexs.add((short)mesh.triItem.get(i).t0);
            objData.indexs.add((short)mesh.triItem.get(i).t1);
            objData.indexs.add((short)mesh.triItem.get(i).t2);
        }
        objData.bindPosAry = bindPosAry;
        objData.invertAry = invertAry;


        return objData;
    }

    private float getW(float x, float y, float z) {
        double t = 1 - (x * x + y * y + z * z);
        if (t < 0) {
            t = 0;
        } else {
           t = -Math.sqrt(t);
        }
        return (float) t;
    }

    private List<ObjectBone> processBoneNew(List<ObjectBone> targetAry) {
        List<ObjectBone> newTargetAry = MeshToObjUtils.getStorNewTargerArr(targetAry);
        //添加bip骨骼到新数组
        int index;
        List<Integer> mapkeyAry  = new ArrayList();//新旧ID映射关系
        for (int i = 0; i < targetAry.size(); i++) {
              index = newTargetAry.indexOf(targetAry.get(i));
            mapkeyAry.add(index);
        }
        List<ObjectBone> resultAry = new ArrayList<>();//最终更新的数据
        for (int i = 0; i < newTargetAry.size(); i++) {//数据复制
            ObjectBone $kkkk = newTargetAry.get(i);
            resultAry.add($kkkk.clone());
        }

        for (int i = 0; i < resultAry.size(); i++) {//从映射关系更新父级id
            index = resultAry.get(i).father;
            if (index != -1) {
                resultAry.get(i).father = mapkeyAry.get(i);
            }
        }

        return resultAry;
    }
}

File: metalios/iosgl/iosgl/pan3d/md5/MeshToObjUtils.m
//
//  MeshToObjUtils.m
//  iosgl
//
//  Created by pan3dme on 2020/11/12.
//  Copyright © 2020 zhao. All rights reserved.
//

#import "MeshToObjUtils.h"
#import "TimeUtil.h"
#import "ObjectUv.h"
#import "ObjectTri.h"
#import "Vector3D.h"
#import "MeshItem.h"
#import "ObjectBone.h"
#import "Quaternion.h"
#import "ObjectWeight.h"
#import "Md5MeshData.h"
#import "AnimManager.h"
 
@implementation MeshToObjUtils
 

-(NSArray*)getStorNewTargerArr:(NSArray*)targetAry
{
    NSMutableArray* newTargetAry  = [[NSMutableArray alloc] init];
    for (int i = 0; i < targetAry.count; i++) {
       NSUInteger idx= [newTargetAry indexOfObject: targetAry[i]];
        if ( idx>newTargetAry.count) {
            [newTargetAry addObject:targetAry[i]];
        }
    }
    return newTargetAry;
 
}
-(void)getObj:(Md5MeshData*)mesh;
{
    MeshData* objData  =mesh;
             objData.vertices = [[NSMutableArray alloc]init];
    objData.uvs = [[NSMutableArray alloc]init];
    objData.nrms = [[NSMutableArray alloc]init];
    objData.indexs =[[NSMutableArray alloc]init];

    NSMutableArray<Matrix3D*>*  bindPosAry  = [[NSMutableArray alloc]init];
    NSMutableArray<Matrix3D*>* invertAry = [[NSMutableArray alloc]init];
    NSMutableArray<MeshItem*>* meshItemAry = [[NSMutableArray alloc]init];
    NSArray* boneItemAry = [self processBoneNew:mesh.boneItem];
    
    for (int i = 0; i < boneItemAry.count; i++) {
        ObjectBone* objbone = boneItemAry[i];
        Quaternion* OldQ =   [[Quaternion alloc]x:objbone.qx y:objbone.qy z:objbone.qz];
        
        OldQ.w = [self.scene3D.animManager getW:OldQ.x  y:OldQ.y z:OldQ.z];
        Matrix3D* newM = [OldQ toMatrix3D];
        [newM appendTranslation:objbone.tx y:objbone.ty z:objbone.tz];
        objbone.matrix = newM;
        [bindPosAry addObject:newM];
        Matrix3D* inverMatrix = [newM clone];
         [inverMatrix Invert];;
        [invertAry addObject:inverMatrix];;
    }
    NSMutableArray* vertices=[[NSMutableArray alloc]init ];
    NSMutableArray* uvs=[[NSMutableArray alloc]init ];
    for (int i = 0; i < mesh.uvItem.count; i++) {
        ObjectUv* objuv = mesh.uvItem[i];
        Vector3D* v3d = [[Vector3D alloc] init];
        NSMutableArray* wAry  = [[NSMutableArray alloc]init];
        for (int j = 0; j < objuv.b; j++) {
            int weightID = objuv.a + j;
            ObjectWeight* objWeight = mesh.weightItem[weightID];
            Matrix3D* ma = ((ObjectBone*)( boneItemAry[objWeight.boneId])).matrix;
            Vector3D* tempV3d =[[Vector3D alloc]x:objWeight.x y:objWeight.y z:objWeight.z];
            tempV3d =[ma transformVector:tempV3d];
            [tempV3d scaleBy:objWeight.w];
            v3d= [v3d add:tempV3d];
            [wAry addObject:[NSNumber numberWithFloat:objWeight.w]];
        }
        [vertices addObject:[NSNumber numberWithFloat: v3d.x]];
        [vertices addObject:[NSNumber numberWithFloat: v3d.y]];
        [vertices addObject:[NSNumber numberWithFloat: v3d.z]];
        [uvs addObject:[NSNumber numberWithFloat: objuv.x]];
        [uvs addObject:[NSNumber numberWithFloat: objuv.y]];
   
        MeshItem* meshitem = [[MeshItem alloc]init];
        meshitem.verts = v3d;
        meshitem.uvInfo = objuv;
        [meshItemAry addObject:meshitem];
 
    }
    NSMutableArray* indexs=[[NSMutableArray alloc]init];
    for (int i = 0; i < mesh.triItem.count; i++) {
        ObjectTri* objectTri=mesh.triItem[i];
        [indexs addObject:[NSNumber numberWithInt:objectTri.t0]];
        [indexs addObject:[NSNumber numberWithInt:objectTri.t1]];
        [indexs addObject:[NSNumber numberWithInt:objectTri.t2]];
    }
    objData.vertices=vertices;
    objData.uvs=uvs;
    objData.indexs=indexs;
    objData.bindPosAry=bindPosAry;
    objData.invertAry=invertAry;
 
 
}
-(int)indexOf:(NSArray*)arr find:(NSObject*)find
{
    NSInteger idx=[arr indexOfObject:find];
    if(idx>arr.count){
        return -1;
    }else{
        return (int)idx;
    }
}
-(NSArray*) processBoneNew:(NSArray* ) targetAry{
    NSArray* newTargetAry=[self.scene3D.meshToObjUtils getStorNewTargerArr: targetAry ];
    NSMutableArray* mapkeyAry  =[[NSMutableArray alloc] init];//新旧ID映射关系
    for (int i = 0; i < targetAry.count; i++) {
        int index=[self indexOf:newTargetAry find:targetAry[i]];
        [mapkeyAry addObject:[NSNumber numberWithInt:index]];
    }
    NSMutableArray* resultAry   = [[NSMutableArray alloc] init];//最终更新的数据
    for (int i = 0; i < newTargetAry.count; i++) {//数据复制
        ObjectBone* objectBone = newTargetAry[i];
        [resultAry addObject:[objectBone clone] ];
    }
    for (int i = 0; i < resultAry.count; i++) {
        ObjectBone* objectBone = resultAry[i] ;
        int index = objectBone.father;
        if (index != -1) {
            objectBone.father = [mapkeyAry[index] intValue];
        }
    }
    return resultAry;
    
}
@end
